okio-files/src/commonMain/kotlin/okio/Path.kt

```diff
@@ -0,0 +1,152 @@
+ * characters. Segments satisfy these rules:
+ *
+ *  * Segments are always non-empty.
+ *  * Segments may not be `.`.
+ *  * Segments with the value `..` only occur in relative paths, where these must precede all other
+ *    segments.
+ *
+ * The only path that ends with `/` is the filesystem root, `/`. The empty path `""` is a relative
+ * path that resolves to whichever path it is resolved against.
+ *
+ * Sample Paths
+ * ------------
+ *
+ * | Path                   | Type       | Parent              |
+ * | :--------------------- | :--------- | :------------------ |
+ * | `/Users/jessewilson`   | Absolute   | `/Users`            |
+ * | `/Users`               | Absolute   | `/`                 |
+ * | `/`                    | Absolute   | null                |
+ * | `src/main/java`        | Relative   | `src/main`          |
+ * | `src/main`             | Relative   | `src`               |
+ * | `src`                  | Relative   | (empty)             |
+ * | (empty)                | Relative   | null                |
+ * | `../../src/main/java`  | Relative   | ` ../../src/main`   |
+ * | `../../src/main`       | Relative   | ` ../../src`        |
+ * | `../../src`            | Relative   | ` ../..`            |
+ * | `../..`                | Relative   | null                |
+ */
+class Path private constructor(
+  private val bytes: ByteString
+) {
+  val isAbsolute: Boolean
+    get() = bytes.startsWith(SLASH)
+
+  val isRelative: Boolean
+    get() = !isAbsolute
+
+  /**
+   * Returns the path immediately enclosing this path. This returns null if this is either the
+   * filesystem root (`/`), the empty path, or a traversal up from the empty path.
+   */
+  val parent: Path?
+    get() {
+      if (bytes == EMPTY || bytes == SLASH || bytes.endsWith(SLASH_DOT_DOT) || bytes == DOT_DOT) {
+        return null // Terminal path.
+      }
+      return when (val lastSlash = bytes.lastIndexOf(SLASH)) {
+        -1 -> Path(EMPTY) // Parent is the current working directory.
+        0 -> Path(bytes.substring(endIndex = 1)) // Parent is the filesystem root '/'.
+        else -> Path(bytes.substring(endIndex = lastSlash))
+      }
+    }
+
+  /** Returns a path that resolves [child] relative to this path. */
+  operator fun div(child: String): Path {
+    val buffer = Buffer()
+    buffer.write(bytes)
+    if (buffer.size > 0) {
+      buffer.write(SLASH)
+    }
+    buffer.writeUtf8(child)
+    return buffer.toPath()
+  }
+
+  override fun equals(other: Any?) = other is Path && other.bytes == bytes
+
+  override fun hashCode() = bytes.hashCode()
+
+  override fun toString() = bytes.utf8()
+
+  companion object {
+    private val SLASH = "/".encodeUtf8()
+    private val DOT = ".".encodeUtf8()
+    private val DOT_DOT = "..".encodeUtf8()
+    private val SLASH_DOT_DOT = "/..".encodeUtf8()
+
+    fun String.toPath(): Path = Buffer().writeUtf8(this).toPath()
```

Comment:
How do I create a path that is not valid UTF-8?

Reply:
API forthcoming. I wanna cover as much happy path as I can, especially into native code, before handling the long tail of gotchas with these things.